src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.cs

```diff
@@ -716,5 +717,230 @@ internal static unsafe ImmutableArray<byte> DecodeEmbeddedPortablePdbDebugDirect
+            try
+            {
+                peImageDirectoryOpt = peImagePath != null ? Path.GetDirectoryName(peImagePath) : null;
+            }
+            catch (Exception e)
+            {
+                throw new ArgumentException(e.Message, nameof(peImagePath), e);
+            }
+
+            DebugDirectoryEntry codeViewEntry, embeddedPdbEntry;
+            ReadPortableDebugTableEntries(out codeViewEntry, out embeddedPdbEntry);
+
+            // First try .pdb file specified in CodeView data (we prefer .pdb file on disk over embedded PDB
+            // since embedded PDB needs decompression which is less efficient than memory-mapping the file).
+            if (pdbFileStreamProvider != null && codeViewEntry.DataSize != 0)
+            {
+                var provider = OpenCodeViewPortablePdb(codeViewEntry, peImageDirectoryOpt, pdbFileStreamProvider, ref errorToReport);
+                if (provider != null)
+                {
+                    return provider;
+                }
+            }
+
+            // if it failed try Embedded Portable PDB (if available):
+            if (embeddedPdbEntry.DataSize != 0)
+            {
+                return OpenEmbeddedPortablePdb(embeddedPdbEntry, ref errorToReport);
+            }
+
+            // Report any metadata and IO errors. PDB migth exist but we couldn't read some metadata. 
+            // The caller might chose to ignore the failure or report it to the user.
+            if (errorToReport != null)
+            {
+                Debug.Assert(errorToReport is BadImageFormatException || errorToReport is IOException);
+                ExceptionDispatchInfo.Capture(errorToReport).Throw();
+            }
+
+            return null;
+        }
+
+        private void ReadPortableDebugTableEntries(out DebugDirectoryEntry codeViewEntry, out DebugDirectoryEntry embeddedPdbEntry)
+        {
+            codeViewEntry = default(DebugDirectoryEntry);
+            embeddedPdbEntry = default(DebugDirectoryEntry);
+
+            foreach (DebugDirectoryEntry entry in ReadDebugDirectory())
+            {
+                if (entry.IsPortableCodeView)
+                {
+                    codeViewEntry = entry;
+                }
+                else if (entry.Type == DebugDirectoryEntryType.EmbeddedPortablePdb)
+                {
+                    embeddedPdbEntry = entry;
+                }
+            }
+        }
+
+        private MetadataReaderProvider OpenCodeViewPortablePdb(DebugDirectoryEntry codeViewEntry, string peImageDirectoryOpt, Func<string, Stream> pdbFileStreamProvider, ref Exception errorToReport)
+        {
+            CodeViewDebugDirectoryData data;
+
+            try
+            {
+                data = ReadCodeViewDebugDirectoryData(codeViewEntry);
+            }
+            catch (Exception e) when (e is BadImageFormatException || e is IOException)
+            {
+                errorToReport = errorToReport ?? e;
```

Comment:
Consider AggregateException instead of dropping second exception so that full info of what was attempted and failed is available for diagnostics.